dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java

```diff
@@ -360,4 +416,210 @@ public UUID toUUID(WorkerIdentity identity) throws IllegalArgumentException {
+    }
+  }
+
+  /**
+   * JSON serializer for Jackson.
+   */
+  public static class JacksonSerializer
+      extends com.fasterxml.jackson.databind.JsonSerializer<WorkerIdentity> {
+    public static final JacksonSerializer INSTANCE = new JacksonSerializer();
+
+    private JacksonSerializer() {
+      // prevent instantiation
+    }
+
+    @Override
+    public void serialize(WorkerIdentity value, JsonGenerator gen, SerializerProvider serializers)
+        throws IOException {
+      gen.writeStartObject();
+      gen.writeNumberField(GsonSerde.FIELD_VERSION, value.mVersion);
+      gen.writeStringField(GsonSerde.FIELD_IDENTIFIER, Hex.encodeHexString(value.mId));
+      gen.writeEndObject();
+    }
+  }
+
+  private static class JacksonDeserializeIntermediate {
+    @JsonProperty(GsonSerde.FIELD_VERSION)
+    @Nullable
+    private final String mIdHexEncoded;
+    @JsonProperty(GsonSerde.FIELD_IDENTIFIER)
+    @Nullable
+    private final Integer mVersion;
+
+    @JsonCreator
+    public JacksonDeserializeIntermediate(
+        @Nullable @JsonProperty(GsonSerde.FIELD_VERSION) Integer version,
+        @Nullable @JsonProperty(GsonSerde.FIELD_IDENTIFIER) String hexEncodedIdentifier) {
+      mVersion = version;
+      mIdHexEncoded = hexEncodedIdentifier;
+    }
+  }
+
+  /**
+   * Deserializer for Jackson.
+   */
+  public static class JacksonDeserializeConverter
+      extends StdConverter<JacksonDeserializeIntermediate, WorkerIdentity> {
+    @Override
+    public WorkerIdentity convert(JacksonDeserializeIntermediate value) {
+      alluxio.grpc.WorkerIdentity.Builder protoBuilder = alluxio.grpc.WorkerIdentity.newBuilder();
+      if (value.mIdHexEncoded != null) {
+        final byte[] identifier;
+        try {
+          identifier = Hex.decodeHex(value.mIdHexEncoded);
+          protoBuilder
+              .setIdentifier(UnsafeByteOperations.unsafeWrap(identifier));
+        } catch (DecoderException e) {
+          throw new IllegalArgumentException(
+              "Malformed hex encoded identifier: " + value.mIdHexEncoded, e);
+        }
+      }
+      if (value.mVersion != null) {
+        protoBuilder.setVersion(value.mVersion);
+      }
+      alluxio.grpc.WorkerIdentity proto = protoBuilder.build();
+      try {
+        return WorkerIdentity.fromProto(proto);
+      } catch (ProtoParsingException e) {
+        throw new IllegalArgumentException(
+            "Invalid worker identity: " + proto, e);
+      }
+    }
+  }
+
+  /**
+   * Funnel to serialize an instance of {@link WorkerIdentity} for hashing.
+   */
+  @SuppressWarnings("UnstableApiUsage")
+  public enum HashFunnel implements Funnel<WorkerIdentity> {
+    INSTANCE;
+
+    @Override
+    // @Nonnull annotation is needed to work around a bug in spotbugs
+    // see https://github.com/spotbugs/spotbugs/pull/2502
+    public void funnel(@Nonnull WorkerIdentity from, PrimitiveSink into) {
+      into.putBytes(from.mId)
+          .putInt(from.mVersion);
```

Comment:
This is mostly a performance optimization. Previously we are using strings of worker's network address as the source of the hash. To turn those into a hash code, you need to first create the string from a `WorkerNetAddress`, and encode the string using UTF-8, and then feed it to the hash function (see [code](https://github.com/Alluxio/alluxio/blob/7b37a862e7bb811c368f1ffb3af6387354f9d9b3/dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java#L246-L247)). This bad for performance, especially considering that you need to repeat the process for as many times as there are virtual nodes of workers.

Switching to the `Hasher` and `Funnel` interfaces avoids the creating of intermediate strings as well as the cost of encoding with UTF-8. 

Personally I don't worry too much about stability issues, as the funnel interface has been around since Guava 11.0. There is a stabilization request on that: https://github.com/google/guava/issues/5863.

Reply:
understood, then lets use this.